parachain/primitives/ethereum/src/lib.rs

```diff
@@ -99,6 +102,32 @@ impl Header {
 		keccak_256(&self.rlp(false)).into()
 	}
 
+	pub fn check_receipt_proof(&self, proof: &[Vec<u8>]) -> Option<Receipt> {
+		match self.apply_merkle_proof(proof) {
+			Some((root, data)) if root == self.receipts_root => rlp::decode(&data).ok(),
+			Some((_, _)) => None,
+			None => None,
+		}
+	}
+
+	pub fn apply_merkle_proof(&self, proof: &[Vec<u8>]) -> Option<(H256, Vec<u8>)> {
+		let mut iter = proof.iter().rev();
+		let bytes = match iter.next() {
+			Some(b) => b,
+			None => return None,
+		};
+		let item_to_prove: mpt::ShortNode = rlp::decode(bytes).ok()?;
+	
+		let final_hash: Option<[u8; 32]> = iter.fold(Some(keccak_256(bytes)), |maybe_hash, bytes| {
+			let expected_hash = maybe_hash?;
+			let node: mpt::FullNode = rlp::decode(bytes).ok()?;
+			let found_hash = node.children.iter().find(|&&hash| Some(expected_hash.into()) == hash);
```

Comment:
if you use `into_iter()` here then there is a chance you can simplify `|&&hash|` to `|&hash|` or  `|hash|`